### PR TITLE
GPII-2660: Initial implementation of "markdown json linter"...

### PR DESCRIPTION
**Pull request by the-t-in-rtf**
See [GPII-2660](https://issues.gpii.net/browse/GPII-2660) for details.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,47 @@
+{
+  "name": "gpii-grunt-mdjson-lint",
+  "description": "A plugin to lint JSON and JSON5 blocks within markdown documents.",
+  "version": "1.0.0",
+  "homepage": "",
+  "author": "Tony Atkins <tony@raisingthefloor.org>",
+  "repository": {
+    "type": "git",
+    "url": "git://gihub.com/the-t-in-rtf/grunt-mdjson-linter.git"
```
Comment (amb26) [inline package.json]: Please update to core repo URL (also note that repo name itself is incorrect)

File: package.json [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  },
+  "licenses": [
+    {
+      "type": "BSD-3-Clause",
+      "url": "/blob/master/LICENSE-BSD-3-Clause"
+    }
+  ],
+  "engines": {
+    "node": ">= 0.8.0"
```
Comment (amb26) [inline package.json]: Probably rash to claim that any of our stuff works with <= node 6 these days given we don't test it

File: src/js/mdjson-linter.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+var mdParser = require("markdown-to-ast");
+var JSON5    = require("json5");
+
+fluid.registerNamespace("gpii.mdjsonLinter");
+
+/**
+ *
+ * @param markdownString {String} - The markdown content to be parsed.
+ * @returns {Array} An array containing any errors detected.
```
Comment (amb26) [inline src/js/mdjson-linter.js]: I think we use @return rather than @returns

File: tasks/mdjsonlint.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                    });
+                }
+            });
+
+            if (errorCount) {
+                grunt.log.error("Found " + errorCount + " errors in " + validPaths.length + " files...");
+                return false;
+            }
+            else {
+                grunt.log.ok("Scanned " + validPaths.length + " files, there were no JSON(5) errors found...");
```
Comment (amb26) [inline tasks/mdjsonlint.js]: Just for consistency, could we make this message in line with all the other lint packages we depend on, which would read "xxx files lint free."

File: Gruntfile.js [hunk 5]
```diff
@@ -0,0 +1,42 @@
+/*
+ * gpii-grunt-mdjson-linter
```
Comment (amb26) [inline Gruntfile.js]: Update for current package name

File: package.json (hunk 1 above)
Comment (the-t-in-rtf) [reply package.json]: It's horribly incorrect, I spelled "gihub" wrong as well.  Anyway, fixed.

File: package.json (hunk 2 above)
Comment (the-t-in-rtf) [reply package.json]: Fixed.

Comment (the-t-in-rtf) [reply package.json]: Oops, I see that I need to go in the other direction.  Actually fixed now.

File: src/js/mdjson-linter.js (hunk 3 above)
Comment (the-t-in-rtf) [reply src/js/mdjson-linter.js]: Fixed.

File: tasks/mdjsonlint.js (hunk 4 above)
Comment (the-t-in-rtf) [reply tasks/mdjsonlint.js]: Done.

File: Gruntfile.js (hunk 5 above)
Comment (the-t-in-rtf) [reply Gruntfile.js]: Fixed here and in the package.json as well.

**Comment by the-t-in-rtf** [comment]
@amb26, as far as I can see I've addressed the feedback for this round, this is ready for further review.

**Comment by the-t-in-rtf** [comment]
In testing another module, I discovered a bug which prevented the plugin from returning a non-zero exit code on errors.  This has been resolved.  @amb26, we should close this out so I can include a full release in the upcoming "rollup" plugin.

**Comment by amb26** [comment]
The dependency markdown-to-ast@4.0.0 has become deprecated - apparently it has now moved to the @textlint namespace https://github.com/textlint/textlint/issues/455

**Comment by the-t-in-rtf** [comment]
@amb26, as we discussed in chat, Grunt prevents me from both upgrading that dependency and using the lint-all plugin (which points to the other dependency).

As far as I can see, we can solve this issue by collapsing this package into gpii-grunt-lint-all.  I just checked npm and my local holdings, the only packages that depend on this one are infusion-docs, the gpii-grunt-lint-all plugin, and a few gpii-X modules of mine that should be updated to use the gpii-grunt-lint-all plugin anyway.  I would propose that we:

1. Review this to our satisfaction and merge.
2. Create a new ticket to cover merging with gpii-grunt-lint-all and review.
3. Cut a new release of gpii-grunt-lint-all.
4. Flag the dev releases of this package as deprecated.
5. I would then submit a raft of small pulls for the affected projects.

**Comment by the-t-in-rtf** [comment]
@amb26, hold off on merging this, I am attempting to unblock the upgrade of ast-markdown per our chats.

**Comment by the-t-in-rtf** [comment]
@amb26, ast-markdown has now been updated and this is ready for further review.

File: Gruntfile.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * Licensed under the BSD-3-Clause license.
+ */
+/* eslint-env node */
+"use strict";
+var fluid = require("infusion");
+var gpii  = fluid.registerNamespace("gpii");
+fluid.require("%gpii-grunt-lint-all/src/load-npm-tasks-properly.js");
+module.exports = function (grunt) {
+    // This is required because of the inherent circularity of this setup.
```
Comment (amb26) [inline Gruntfile.js]: Could you list more clearly here what the circle is and what modules it passes through

File: Gruntfile.js [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    gpii.grunt.lintAll.fixGruntTaskLoading(grunt);
+
+    grunt.initConfig({
+        lintAll: {
+            sources: {
+                md: ["./*.md", "./**/*.md", "!./node_modules/**", "!./tests/markdown/bad*"],
```
Comment (amb26) [inline Gruntfile.js:30]: What did we decide about the possibility of "ignores" that always bypasses top-level node_modules directories?

File: package.json [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        "test": "node node_modules/istanbul/lib/cli.js cover tests/all-tests.js"
+    },
+    "dependencies": {
+        "gpii-grunt-lint-all": "1.0.1-dev.20180706T153657Z.4cbbd61",
+        "infusion": "3.0.0-dev.20171006T195039Z.a128358",
+        "json5": "1.0.1",
+        "@textlint/markdown-to-ast": "6.0.3"
+    },
+    "devDependencies": {
+        "grunt": "~1.0.1",
```
Comment (amb26) [inline package.json]: Please fix this version to the latest

File: Gruntfile.js (hunk 6 above)
Comment (the-t-in-rtf) [reply Gruntfile.js]: Done.

File: package.json (hunk 8 above)
Comment (the-t-in-rtf) [reply package.json]: Done, I had wrong-headedly thought that was indicating compatibility for consumers and wanted to be flexible.

File: Gruntfile.js (hunk 7 above)
Comment (the-t-in-rtf) [reply Gruntfile.js:30]: I had to review to confirm.  The gpii-grunt-lint-all plugin does not use expansion or any other Fluid conventions, but uses grunt's own variable syntax to mash together options for each type of file, as shown here:

https://github.com/GPII/gpii-grunt-lint-all/blob/master/tasks/lint-all.js#L39

The best I could manage without more serious surgery is to add an ignores option, which is included in each check.  The caveat is that those ignores would have to be expressed as negated patterns.  If that's acceptable, I can write up that work against gpii-grunt-lint-all and knock that out in short order.

**Comment by the-t-in-rtf** [comment]
@amb26, I responded to your feedback and updated all dependencies, this should be ready for review again.

File: Gruntfile.js (hunk 7 above)
Comment (the-t-in-rtf) [reply Gruntfile.js:30]: A first pass at global ignores has already been added to the existing gpii-grunt-lint-all pull.

**Comment by the-t-in-rtf** [comment]
I can't cut the tag/release for this one, given I'm the original author, can I get perms similar to what I have for gpii-express.

**Comment by amb26** [comment]
I've added the same team structure to this repo as for gpii-express - let me know if this is now working

**Comment by the-t-in-rtf** [comment]
Thanks, I was able to cut the tug and create the release.

**Comment by the-t-in-rtf** [comment]
I also invited you to be an admin in the new npm package.

**Pull request merged**